From 11f3770adbea24b2d59007f943348249f65fa431 Mon Sep 17 00:00:00 2001
From: Yong Li <yong.li@intel.com>
Date: Mon, 19 Feb 2024 16:29:13 +0800
Subject: [PATCH 56/63] lib: sbi: add support to update domain context entry

Some of the domain os needs to update the context entry
dynamically eg: optee

The patch provides an addition api to support update
the domain context entry.

Signed-off-by: Yong Li <yong.li@intel.com>
---
 include/sbi/sbi_domain_context.h | 10 ++++++++++
 lib/sbi/sbi_domain_context.c     | 20 ++++++++++++++++++++
 2 files changed, 30 insertions(+)

diff --git a/include/sbi/sbi_domain_context.h b/include/sbi/sbi_domain_context.h
index edba764..b0848a0 100755
--- a/include/sbi/sbi_domain_context.h
+++ b/include/sbi/sbi_domain_context.h
@@ -57,6 +57,16 @@ struct sbi_context {
 		sbi_hartid_to_hartindex(current_hartid()), \
 		sbi_domain_thishart_ptr())
 
+/**
+ * Set domain entry point
+ * @param dom pointer to domain
+ * @param entry_point new entry point of domain
+ *
+ * @return 0 on success and negative error code on failure
+ */
+int sbi_domain_context_set_mepc(struct sbi_domain *dom, unsigned long entry_point);
+
+
 /**
  * Enter a specific domain context synchronously
  * @param dom pointer to domain
diff --git a/lib/sbi/sbi_domain_context.c b/lib/sbi/sbi_domain_context.c
index d6843a6..9213c1a 100755
--- a/lib/sbi/sbi_domain_context.c
+++ b/lib/sbi/sbi_domain_context.c
@@ -79,6 +79,26 @@ static void switch_to_next_domain_context(struct sbi_context *ctx,
 	}
 }
 
+/**
+ * Set domain entry point
+ * @param dom pointer to domain
+ * @param entry_point new entry point of domain
+ *
+ * @return 0 on success and negative error code on failure
+ */
+int sbi_domain_context_set_mepc(struct sbi_domain *dom, unsigned long entry_point)
+{
+	struct sbi_context *dom_ctx = sbi_hartindex_to_domain_context(
+		sbi_hartid_to_hartindex(current_hartid()), dom);
+	/* Validate the domain context existence */
+	if (!dom_ctx)
+		return SBI_EINVAL;
+
+	/* Point to the last instruction address */
+	dom_ctx->regs.mepc = entry_point - 4;
+	return 0;
+}
+
 int sbi_domain_context_enter(struct sbi_domain *dom)
 {
 	struct sbi_context *ctx = sbi_domain_context_thishart_ptr();
-- 
2.34.1

